File: coreui/src/main/java/com/mbg/module/ui/mvp/holder/PresenterHolder.java
package com.mbg.module.ui.mvp.holder;

import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.OnLifecycleEvent;


import com.mbg.module.ui.mvp.MvpPresenter;
import com.mbg.module.ui.mvp.base.IView;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;


/**
 * Presenter管理类，用来保存presenter
 */
public final class PresenterHolder {
    private static volatile PresenterHolder holder;
    private final Map<String, MvpPresenter<? extends IView>> pool;//存储Presenter

    private PresenterHolder() {
        pool = new ConcurrentHashMap<>();
    }

    public static PresenterHolder get() {
        if (holder == null) {
            synchronized (PresenterHolder.class) {
                if (holder == null) {
                    holder = new PresenterHolder();
                }
            }
        }

        return holder;
    }


    /**
     * 保存presenter
     *
     * @param presenter p
     * @param lifecycle l
     */
    public void addPresenter(final MvpPresenter<? extends IView> presenter, final Lifecycle lifecycle) {

        if (pool.containsKey(presenter.getPresenterId())) {
            pool.remove(presenter.getPresenterId());
        }
        pool.put(presenter.getPresenterId(), presenter);

        lifecycle.addObserver(new LifecycleObserver() {
            @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
            public void onDestroy() {
                pool.remove(presenter.getPresenterId());
            }
        });

    }

    /**
     * 获取presenter
     *
     * @param presenterId id
     */
    public MvpPresenter<? extends IView> getPresenter(final String presenterId) {
        return pool.get(presenterId);
    }
}

File: coreui/src/main/java/com/mbg/module/ui/mvp/kotlin/holder/PresenterHolder.kt
package com.mbg.module.ui.mvp.kotlin.holder

import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.OnLifecycleEvent
import com.mbg.module.ui.mvp.kotlin.MvpPresenter
import com.mbg.module.ui.mvp.kotlin.base.IntView

import java.util.concurrent.ConcurrentHashMap

class PresenterHolder private constructor(){

    //存储Presenter
    private val pool: MutableMap<String, MvpPresenter<out IntView>> = ConcurrentHashMap()


    /**
     * 保存presenter
     *
     * @param presenter p
     * @param lifecycle l
     */
    fun addPresenter(presenter: MvpPresenter<out IntView>, lifecycle: Lifecycle) {
        if (pool.containsKey(presenter.getPresenterId())) {
            pool.remove(presenter.getPresenterId())
        }

        pool[presenter.getPresenterId()] = presenter
        lifecycle.addObserver(object : LifecycleObserver {
            @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
            fun onDestroy() {
                if (pool.containsKey(presenter.getPresenterId())) {
                    pool.remove(presenter.getPresenterId())
                }
            }
        })
    }

    /**
     * 获取presenter
     *
     * @param presenterId id
     */
    fun getPresenter(presenterId: String): MvpPresenter<out IntView>? {
        return pool[presenterId]
    }

    companion object{
        private val holder:PresenterHolder by lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED) {
            PresenterHolder()
        }

        @JvmStatic
        fun get():PresenterHolder{
            return holder
        }
    }
}